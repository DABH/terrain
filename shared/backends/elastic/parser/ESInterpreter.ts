/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import EQLConfig from './EQLConfig';
import ESJSONParser from './ESJSONParser';
import ESParserError from './ESParserError';
import ESValueInfo from './ESValueInfo';

export const ESInterpreterDefaultConfig = new EQLConfig();

/**
 * An instrumented interpreter that takes the output of ESJSONParser and
 * decorates the results with ES-specific information.
 */
export default class ESInterpreter
{
  public parser: ESJSONParser; // source parser
  public config: EQLConfig; // query language description

  /**
   * Runs the interpreter on the given query string. Read needed data by calling the
   * public member functions below. You can also pass in an existing ESJSONParser
   * to run the interpreter on it's result.
   * @param query the query string or parser to interpret
   */
  public constructor(query: string | ESJSONParser, config: EQLConfig = ESInterpreterDefaultConfig)
  {
    this.config = config;

    if (typeof query === 'string')
    {
      this.parser = new ESJSONParser(query);
    } else
    {
      this.parser = query;
    }

    if (this.parser.getValue() !== null && this.parser.hasError() === false)
    {
      try
      {
        this.config.getClause('root').mark(this, this.parser.getValueInfo());
      } catch (e)
      {
        this.accumulateError(this.parser.getValueInfo(), 'Failed to mark the json object ' + String(e.message));
      }
    }
  }

  public accumulateError(info: ESValueInfo, message: string, isWarning: boolean = false): void
  {
    this.parser.accumulateError(new ESParserError(info.tokens[0], info, message, isWarning));
  }

  public hasError()
  {
    return this.parser === null || this.parser.hasError();
  }
}
