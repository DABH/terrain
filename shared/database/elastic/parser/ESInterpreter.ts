/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import ESParser from 'shared/database/elastic/parser/ESParser';
import ESClause from './clauses/ESClause';
import EQLConfig from './EQLConfig';
import ESJSONParser from './ESJSONParser';
import ESParserError from './ESParserError';
import ESValueInfo from './ESValueInfo';

export const ESInterpreterDefaultConfig = new EQLConfig();

/**
 * An instrumented interpreter that takes the output of ESJSONParser and
 * decorates the results with ES-specific information.
 */
export default class ESInterpreter
{
  public config: EQLConfig; // query language description
  public params: { [name: string]: null | ESClause }; // input parameter clause types
  public parser: ESParser; // source parser
  public rootValueInfo: ESValueInfo;
  public errors: ESParserError[];

  /**
   * Runs the interpreter on the given query string. Read needed data by calling the
   * public member functions below. You can also pass in an existing ESJSONParser
   * to run the interpreter on it's result.
   *
   * 1) parse
   * 2) interpret
   *
   * @param query the query string or parser to interpret
   * @param config the spec config to use
   * @param params parameter map to use
   */
  public constructor(query: string | ESParser,
    params: { [name: string]: any } = {},
    config: EQLConfig = ESInterpreterDefaultConfig)
  {
    this.config = config;
    this.params = params;
    this.errors = [];

    if (typeof query === 'string')
    {
      this.parser = new ESJSONParser(query) as ESParser;
      if (this.parser.hasError())
      {
        this.accumulateError(null, 'Failed to parse the query ' + query);
        return;
      }
      this.rootValueInfo = this.parser.getValueInfo();
    } else if (query instanceof ESParser)
    {
      this.parser = query;
      this.rootValueInfo = this.parser.getValueInfo();
    } else
    {
      this.parser = null;
      this.accumulateError(null, 'The input must be a query, an ESJSONParser object, or an ESValueInfo');
      return;
    }

    try
    {
      const root: ESValueInfo = this.rootValueInfo;
      if (root.clause === undefined)
      {
        root.clause = this.config.getClause('root');
      }
      root.recursivelyVisit(
        (info: ESValueInfo): boolean =>
        {
          if (info.parameter !== undefined)
          {
            const value: null | any = this.params[info.parameter];
            if (value === undefined)
            {
              this.accumulateError(info, 'Undefined parameter: ' + info.parameter);
            }

            return false; // don't validate parameters
          }

          if (info.clause !== undefined)
          {
            info.clause.mark(this, info);
          }

          return true;
        });
    } catch (e)
    {
      this.accumulateError(this.rootValueInfo, 'Failed to mark the json object ' + String(e.message));
    }
  }

  public accumulateError(info: ESValueInfo, message: string, isWarning: boolean = false): void
  {
    let token = null;
    if (info !== null && info !== undefined && info.tokens.length > 0)
    {
      token = info.tokens[0];
    }
    const e = new ESParserError(token, info, message, isWarning);
    this.errors.push(e);
    if (this.parser !== null)
    {
      this.parser.accumulateError(e);
    }
  }

  public hasInterpretingError()
  {
    return this.errors.length > 0;
  }

  public hasJSONParsingError()
  {
    return this.parser === null || this.parser.hasError();
  }

  public hasError()
  {
    return this.hasInterpretingError() || this.hasJSONParsingError();
  }
}
