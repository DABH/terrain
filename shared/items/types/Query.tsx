/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

import {Cards} from '../../blocks/types/Card';
import BlockUtils from '../../blocks/BlockUtils';

// A query can be viewed and edited in the Builder
// currently, only Variants have Queries, 1:1, but that may change
class QueryC
{
  type: 'QUERY' = 'QUERY';
  parent: number = -1;
  name: string = '';
  status: 'BUILD' | 'LIVE' = 'BUILD';
  language: 'elastic';
  
  id: ID = -1;
  variantId: number = -1;
  
  // TODO change?
  db: {
    id: ID;
    name: string;
    source: 'm1' | 'm2';
    type: string;
  } = {} as any;

  cards: Cards = List([]);
  inputs: List<any> = List([]);
  resultsConfig: IResultsConfig = null;
  tql: string = '';
  deckOpen: boolean = true;

  cardsAndCodeInSync: boolean = false;
  parseError: string = null;
  
  
  dbFields = ['id', 'parent', 'name', 'status', 'type'];
  excludeFields= ['dbFields', 'excludeFields'];
  
  modelVersion = 2; // 2 is for the first version of Node midway
}
const Query_Record = Immutable.Record(new QueryC());
export interface Query extends QueryC, IRecord<Query> {}

export const _Query = (config?: Object) => {
  config = config || {};
  config['cards'] = BlockUtils.recordFromJS(config['cards'] || []);
  config['inputs'] = BlockUtils.recordFromJS(config['inputs'] || []);
  config['resultsConfig'] = _IResultsConfig(config['resultsConfig']);

  let query = new Query_Record(config) as any as Query;

  return query;
};

export function queryForSave(query: Query): Object
{
  query = query
    .set('cards', cardsForServer(query.cards))
    .set('resultsConfig', query.resultsConfig.toJS());
  return query.toJS();
}

export default Query;
