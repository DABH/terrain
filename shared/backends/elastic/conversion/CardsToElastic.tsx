/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

// tslint:disable:restrict-plus-operands strict-boolean-expressions

import * as Immutable from 'immutable';
import * as _ from 'underscore';
import * as CommonElastic from '../syntax/CommonElastic';

import * as BlockUtils from '../../../blocks/BlockUtils';
import { Block, TQLRecursiveObjectFn } from '../../../blocks/types/Block';
import { Card } from '../../../blocks/types/Card';
import { Input, InputPrefix, InputType } from '../../../blocks/types/Input';
import Query from '../../../items/types/Query';
import ElasticBlocks from '../blocks/ElasticBlocks';

import ESParameterFiller from '../parser/EQLParameterFiller';
import ESParser from '../parser/ESJSONParser';
import ESParserError from '../parser/ESParserError';
import ESValueInfo from '../parser/ESValueInfo';
import ESConverter from './formatter/ESConverter';

const join = (j, index) => (index === 0 ? '' : j);
const addTabs = (str) => ' ' + str.replace(/\n/g, '\n ');
const removeBlanks = (str) => str.replace(/\n[ \t]*\n/g, '\n');
type PatternFn = (obj: any, index?: number, isLast?: boolean) => string;

export interface Options
{
  allFields?: boolean; // amend the final Select card to include all possible fields.
  limit?: number;
  count?: boolean;
  transformAliases?: boolean; // if true, scan the top Select for Transforms, and add an alias row using the transform's ID
  replaceInputs?: boolean; // replaces occurences of inputs with their values
}

export interface ElasticObjectInterface
{
  index?: string;
  type?: string;
  body?: {
    _source: object;
  };

  [key: string]: any;
}

export function isInput(name: string, inputs: Immutable.List<Input>)
{
  return inputs && name.charAt(0) === InputPrefix &&
    (inputs.findIndex((input: Input) => (name.substring(1) === input.key)) > -1);
}

export function toInputMap(inputs: Immutable.List<Input>): object
{
  const inputMap: object = {};
  inputs.map((input: Input) =>
  {
    inputMap[input.key] = JSON.parse(input.value);
  });
  return inputMap;
}

export function stringifyWithParameters(
  obj: object | number | boolean | string | null,
  inputs?: Immutable.List<Input>): string | null
{
  if (typeof obj === 'number' || typeof obj === 'boolean' || obj === null)
  {
    return '' + obj;
  }
  else if (typeof obj === 'string')
  {
    if (isInput(obj, inputs))
    {
      return obj;
    }
    return '"' + obj + '"';
  }
  else if (typeof obj === 'object')
  {
    let str = '{';
    const keys = Object.keys(obj);
    for (let i = 0; i < keys.length; i++)
    {
      str += '"' + keys[i] + '": ';
      str += stringifyWithParameters(obj[keys[i]], inputs);
      if (i < keys.length - 1)
      {
        str += ',';
      }
    }
    str += '}';
    return str;
  }
  else
  {
    return '';
  }
}

class CardsToElastic
{
  public static toElastic(query: Query, options: Options = {}): string
  {
    const elasticObj: ElasticObjectInterface = {};

    query.cards.map(
      (card: Card) =>
      {
        const val = CardsToElastic.blockToElastic(card, options);
        const key = card['key'];

        if (key)
        {
          elasticObj[key] = val;
        }
        else if (typeof val === 'object' && !Array.isArray(val))
        {
          _.extend(elasticObj, val);
        }
      },
    );

    if (options.allFields === true)
    {
      if (elasticObj.body && elasticObj.body._source)
      {
        elasticObj.body._source = [];
      }
    }

    const text: string = stringifyWithParameters(elasticObj, query.inputs);
    const parser: ESParser = new ESParser(text, true);
    if (options.replaceInputs)
    {
      const valueInfo: ESValueInfo = parser.getValueInfo();
      const params = toInputMap(query.inputs);
      const result = ESParameterFiller.generate(valueInfo, params);
      return ESConverter.formatES(new ESParser(result));
    }
    else
    {
      // TODO: pipe this through the formatter once it can handle parameters
      // return ESConverter.formatES(new ESParser(text));
      return text;
    }
  }

  public static blockToElastic(block: Block, options: Options = {}): string | object | number | boolean
  {
    if (typeof block !== 'object')
    {
      return block;
    }

    if (block && block.static.tql)
    {
      if (typeof block['value'] === 'string' && block['value'].charAt(0) === InputPrefix)
      {
        return block['value'];
      }
      const tql = block.static.tql as TQLRecursiveObjectFn;
      return tql(block, CardsToElastic.blockToElastic, options);
    }
    return { notYet: 'not yet done' };
  }
}

export default CardsToElastic;
