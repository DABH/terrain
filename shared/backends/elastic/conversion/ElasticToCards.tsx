/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as Immutable from 'immutable';
import * as _ from 'underscore';
import List = Immutable.List;
import Map = Immutable.Map;

import AjaxM1 from '../../../../src/app/util/AjaxM1'; // TODO change / remove
import Query from '../../../items/types/Query';
import * as CommonElastic from '../syntax/CommonElastic';

import BlockUtils from '../../../blocks/BlockUtils';
import { Block } from '../../../blocks/types/Block';
import { Card, Cards, CardString } from '../../../blocks/types/Card';
const { make } = BlockUtils;
import Blocks from '../blocks/ElasticBlocks';

export default function ElasticToCards(
  query: Query,
  queryReady: (query: Query) => void,
): Query
{
  try
  {
    const obj = JSON.parse(query.tql);
    let cards = parseMagicObject(obj);
    cards = BlockUtils.reconcileCards(query.cards, cards);
    return query
      .set('cards', cards)
      .set('cardsAndCodeInSync', true);
  }
  catch (e)
  {
    return query
      .set('cardsAndCodeInSync', false);
  }
}

const parseObjectWrap = (obj: object): Cards =>
{
  const arr: Card[] = _.map(obj,
    (value: any, key: string) =>
    {
      return make(
        Blocks.elasticKeyValueWrap,
        {
          key,
          cards: Immutable.List([
            parseValueSingleCard(value),
          ]),
        },
      );
    },
  );

  return Immutable.List(arr);
};

const parseArrayWrap = (arr: any[]): Cards =>
{
  return Immutable.List(arr.map(parseValueSingleCard));
};

const parseValueSingleCard = (value: any): Card =>
{
  switch (typeof value)
  {
    case 'string':
      return make(Blocks.elasticText, {
        value,
      });
    case 'number':
      return make(Blocks.elasticNumber, {
        value,
      });
    case 'boolean':
      return make(Blocks.elasticBool, {
        value: value ? 1 : 0,
      });
    case 'object':
      if (value === null)
      {
        return make(Blocks.elasticNull);
      }
      else if (Array.isArray(value))
      {
        return make(
          Blocks.elasticArray,
          {
            cards: parseArrayWrap(value),
          },
        );
      }
      else
      {
        // value is an object
        return make(
          Blocks.elasticObject,
          {
            cards: parseObjectWrap(value),
          },
        );
      }
    default:
      throw new Error('Elastic Parsing: Unsupported value: ' + value);
  }
};

const parseMagicObject = (obj: object): Cards =>
{
  const values: Card[] = _.map(obj,
    (value: any, key: string) =>
    {
      if (Array.isArray(value))
      {
        return make(Blocks.elasticMagicList, {
          values: parseMagicObject(value),
        });
      }
      else if (typeof value === 'object')
      {
        const cards: Cards = parseMagicObject(value);
        return make(Blocks.elasticMagicValue, {
          key,
          value: cards.first(),
        });
      }
      else
      {
        return make(Blocks.elasticMagicValue, {
          key,
          value,
        });
      }
    },
  );

  const magicCard = make(
    Blocks.elasticMagicCard,
    {
      values: Immutable.List(values),
    },
  );

  return Immutable.List([magicCard]);
};
