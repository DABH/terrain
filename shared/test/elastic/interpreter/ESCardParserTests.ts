/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as fs from 'fs';
import ESInterpreter from 'shared/database/elastic/parser/ESInterpreter';
import * as winston from 'winston';
import ESJSONParser from '../../../database/elastic/parser/ESJSONParser';
import { makePromiseCallback } from '../../Utils';

// make sure importing ESCardParser before importing ElasticToCards
import ESCardParser from 'src/database/elastic/conversion/ESCardParser';

import { ElasticValueInfoToCards, parseCardFromValueInfo } from 'src/database/elastic/conversion/ElasticToCards';

import * as Immutable from 'immutable';
import ESParserError from 'shared/database/elastic/parser/ESParserError';
import CardsToElastic from 'src/database/elastic/conversion/CardsToElastic';

import Query, { _Query } from 'src/items/types/Query';

function getExpectedFile(): string
{
  return __filename.split('.')[0] + '.expected';
}

let expected;

beforeAll(async (done) =>
{
  // TODO: get rid of this monstrosity once @types/winston is updated.
  (winston as any).level = 'debug';

  const contents: any = await new Promise((resolve, reject) =>
  {
    fs.readFile(getExpectedFile(), makePromiseCallback(resolve, reject));
  });

  expected = JSON.parse(contents);
  done();
});

function testCardParse(testName: string,
  testString: string,
  expectedValue: any,
  expectedErrors: ESParserError[] = [])
{
  winston.info('testing "' + testName + '": "' + testString + '"');
  const emptyCards = Immutable.List([]);
  const interpreter: ESInterpreter = new ESInterpreter(testString);
  const parser: ESJSONParser = interpreter.parser as ESJSONParser;
  const rootValueInfo = parser.getValueInfo();
  const rootCards = ElasticValueInfoToCards(rootValueInfo, Immutable.List([]), _Query());
  // parse the card
  const rootCard = rootCards.get(0);
  expect(rootCard['type']).toEqual('eqlbody');
  const cardParser = new ESCardParser(rootCard);
  // interpreting the parsed card
  const cardInterpreter = new ESInterpreter(cardParser);
  expect(cardInterpreter.errors).toEqual(expectedErrors);
  expect(CardsToElastic.blockToElastic(rootCard, {}, _Query())).toEqual(expectedValue);
}

test('parse card', () =>
{
  Object.getOwnPropertyNames(expected).forEach(
    (testName: string) =>
    {
      const testValue: any = expected[testName];
      testCardParse('test', JSON.stringify(testValue), testValue);
    });
});
