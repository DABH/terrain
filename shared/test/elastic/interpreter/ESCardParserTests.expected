{
  "empty query" : {
  },
  "simple query" : {
    "from" : 0,
    "size" : 10
  },
  "nested simple query" : {
      "from" : 0,
      "size" : 10,
      "query" : {
        "bool" : {
          "must" : [
            {"match" : {"_index" : "movies"}},
            {"match" : {"_type" : "data"}}
          ]
        }
      }
    }
}
